src/lib.rs

```diff
@@ -104,293 +235,367 @@ fn serialize_record_batch(batch: RecordBatch) -> Result<ArrowBatch> {
-                return IcebergResult::IcebergError;
-            }
-        }
-    };
+pub extern "C" fn iceberg_init_runtime(
+    config: IcebergStaticConfig,
+    panic_callback: PanicCallback,
+    result_callback: ResultCallback,
+) -> CResult {
+    // Set the result callback
+    if let Err(_) = RESULT_CB.set(result_callback) {
+        return CResult::Error; // Already initialized
+    }
 
-    let metadata_path_str = unsafe {
-        match CStr::from_ptr(metadata_path).to_str() {
-            Ok(s) => s,
-            Err(e) => {
-                set_error(format!("Invalid UTF-8 in metadata path: {}", e));
-                return IcebergResult::IcebergError;
-            }
+    // Set up panic hook
+    let prev = std::panic::take_hook();
+    std::panic::set_hook(Box::new(move |info| {
+        prev(info);
+        unsafe { panic_callback() };
+    }));
+
+    // Set up logging if not already configured
+    if std::env::var("RUST_LOG").is_err() {
+        unsafe { std::env::set_var("RUST_LOG", "iceberg_rust_ffi=warn,iceberg=warn") }
+    }
+
+    // Initialize tracing subscriber
+    let _ = tracing_subscriber::fmt::try_init();
+
+    // Build tokio runtime
+    let mut rt_builder = tokio::runtime::Builder::new_multi_thread();
+    rt_builder.enable_all();
+
+    // Configure Julia thread adoption for Julia integration
+    rt_builder.on_thread_start(|| {
+        #[cfg(feature = "julia")]
+        {
+            unsafe { jl_adopt_thread() };
+            unsafe { jl_gc_safe_enter() };
+            unsafe { jl_gc_disable_finalizers_internal() };
         }
+    });
+
+    if config.n_threads > 0 {
+        rt_builder.worker_threads(config.n_threads);
+    }
+
+    let runtime = match rt_builder.build() {
+        Ok(rt) => rt,
+        Err(_) => return CResult::Error,
     };
 
-    // TODO: Perhaps we should have full asynchronicity that includes the caller code (e.g. Julia) instead of blocking here.
-    let result: Result<iceberg::table::Table, anyhow::Error> = get_runtime().block_on(async {
-        // println!("DEBUG: Table path: {}", path_str);
-        // println!("DEBUG: Metadata path: {}", metadata_path_str);
+    if RT.set(runtime).is_err() {
+        return CResult::Error;
+    }
+
+    CResult::Ok
+}
+
+// Use export_runtime_op! macro for table opening
+export_runtime_op!(
+    iceberg_table_open,
+    IcebergTableResponse,
+    || {
+        let table_path_str = unsafe {
+            CStr::from_ptr(table_path).to_str()
+                .map_err(|e| anyhow::anyhow!("Invalid UTF-8 in table path: {}", e))?
+        };
+        let metadata_path_str = unsafe {
+            CStr::from_ptr(metadata_path).to_str()
+                .map_err(|e| anyhow::anyhow!("Invalid UTF-8 in metadata path: {}", e))?
+        };
+        Ok((table_path_str.to_string(), metadata_path_str.to_string()))
+    },
+    paths,
+    async {
+        let (table_path_str, metadata_path_str) = paths;
```

Comment:
I'm not sure if it was a good idea to split table path and relative metadata path, instead of just providing a full path. Perhaps we can wrap the builder code and the async code in functions, and expose both variants simply by having 2 macro invocations?